dialog/file.go

```diff
@@ -0,0 +1,244 @@
+	favourites := widget.NewGroup("Favourites", f.loadFavourites()...)
+	return fyne.NewContainerWithLayout(layout.NewBorderLayout(header, footer, favourites, nil),
+		favourites, header, footer, body)
+}
+
+func (f *fileDialog) loadFavourites() []fyne.CanvasObject {
+	home, _ := os.UserHomeDir()
+	places := []fyne.CanvasObject{
+		widget.NewButton("Home", func() {
+			f.setDirectory(home)
+		}),
+		widget.NewButton("Documents", func() {
+			f.setDirectory(filepath.Join(home, "Documents"))
+		}),
+		widget.NewButton("Downloads", func() {
+			f.setDirectory(filepath.Join(home, "Downloads"))
+		}),
+	}
+
+	places = append(places, f.loadPlaces()...)
+	return places
+}
+
+func (f *fileDialog) setFileDir(dir *fileIcon) {
+	f.setDirectory(dir.path)
+}
+
+func (f *fileDialog) setDirectory(dir string) {
+	f.setFile(nil)
+	f.dir = dir
+
+	f.breadcrumb.Children = nil
+	buildDir := filepath.VolumeName(dir)
+	for i, d := range strings.Split(dir, string(filepath.Separator)) {
+		if d == "" {
+			if i > 0 { // what we get if we split "/"
+				break
+			}
+			buildDir = "/"
+			d = "/"
+		} else if i > 0 {
+			buildDir = filepath.Join(buildDir, d)
+		} else {
+			d = buildDir
+			buildDir = d + string(os.PathSeparator)
+		}
+
+		newDir := buildDir
+		f.breadcrumb.Append(
+			widget.NewButton(d, func() {
+				f.setDirectory(newDir)
+			}),
+		)
+	}
+
+	f.refreshDir(dir)
+}
+
+func (f *fileDialog) refreshDir(dir string) {
+	f.files.Objects = nil
+
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		fyne.LogError("Unable to read path "+dir, err)
+		return
+	}
+
+	var icons []fyne.CanvasObject
+	parent := filepath.Dir(dir)
+	if parent != dir {
+		icons = append(icons, f.newFileIcon(theme.FolderOpenIcon(), filepath.Dir(dir), f.setFileDir))
+	}
+	for _, file := range files {
+		if isHidden(file.Name(), dir) {
+			continue
+		}
+
+		itemPath := filepath.Join(dir, file.Name())
+		if file.IsDir() {
+			icons = append(icons, f.newFileIcon(theme.FolderIcon(), itemPath, f.setFileDir))
+		} else {
+			icons = append(icons, f.newFileIcon(theme.FileIcon(), itemPath, f.setFile))
+		}
+	}
+
+	f.files.Objects = icons
+	f.files.Refresh()
+}
+
+func (f *fileDialog) setFile(file *fileIcon) {
+	if f.current != nil {
+		f.current.current = false
+		f.current.Refresh()
+	}
+	f.current = file
+
+	if file == nil || file.path == "" {
+		f.fileName.SetText("")
+		f.open.Disable()
+	} else {
+		f.fileName.SetText(filepath.Base(file.path))
+		f.open.Enable()
+	}
+}
+
+func showFileDialog(save bool, callback func(string), parent fyne.Window) {
+	if fileOSOverride(save, callback, parent) {
+		return
+	}
+	d := &fileDialog{callback: callback, save: save, parent: parent}
+	ui := d.makeUI()
+	dir, err := os.UserHomeDir()
+	if err != nil {
+		fyne.LogError("Could not load user home dir", err)
+		dir, _ = os.Getwd() //fallback
+	}
+	d.setDirectory(dir)
+
+	spacer := canvas.NewRectangle(theme.BackgroundColor())
+	spacer.SetMinSize(fyne.NewSize(436, 300))
```

Comment:
A fixed min size? Where come these numbers from?

Reply:
A fair question - in my testing it was just the right size to show 4 file icons wide and 2 high.
You're right though this could be calculated. And now I think about it 2.2 or 2.5 high is probably better so you can more easily see to scale!